src/irmin/tree.ml

```diff
@@ -448,79 +446,67 @@ module Make (P : S.PRIVATE) = struct
           v
       | _, v -> v
 
-    let hash_of_value t v =
-      cnt.node_hash <- cnt.node_hash + 1;
-      let k = P.Node.Key.hash v in
-      t.info.hash <- Some k;
-      match t.info.hash with Some k -> k | None -> k
-
-    let hash ~value_of_adds ~value_of_map t =
+    let rec hash : type a. t -> (hash -> a) -> a =
+     fun t k ->
       match cached_hash t with
-      | Some h -> h
+      | Some h -> k h
       | None -> (
-          let of_value v = hash_of_value t v in
+          let a_of_value v =
+            cnt.node_hash <- cnt.node_hash + 1;
+            let h = P.Node.Key.hash v in
+            t.info.hash <- Some h;
+            k h
+          in
           match cached_value t with
-          | Some v -> of_value v
+          | Some v -> a_of_value v
           | None -> (
               match t.v with
-              | Hash (_, h) -> h
-              | Value (_, v, None) -> of_value v
-              | Value (_, v, Some a) -> of_value (value_of_adds t v a)
-              | Map m -> of_value (value_of_map t m)))
+              | Hash (_, h) -> k h
+              | Value (_, v, None) -> a_of_value v
+              | Value (_, v, Some a) -> value_of_adds t v a a_of_value
+              | Map m -> value_of_map t m a_of_value))
 
-    let rec value_of_map t ~value_of_adds map =
+    and value_of_map : type a. t -> map -> (value -> a) -> a =
+     fun t map k ->
       if StepMap.is_empty map then (
         t.info.value <- Some P.Node.Val.empty;
-        P.Node.Val.empty)
+        k P.Node.Val.empty)
       else
         let alist = StepMap.bindings map in
         let rec aux acc = function
           | [] ->
-              let alist = List.rev acc in
               cnt.node_val_v <- cnt.node_val_v + 1;
-              P.Node.Val.v alist
+              let v = P.Node.Val.v (List.rev acc) in
+              t.info.value <- Some v;
+              k v
           | (step, v) :: rest -> (
               match v with
               | `Contents (c, m) ->
                   let v = `Contents (Contents.hash c, m) in
-                  (aux [@tailcall]) ((step, v) :: acc) rest
-              | `Node n ->
-                  let n =
-                    hash ~value_of_adds
-                      ~value_of_map:(value_of_map ~value_of_adds)
-                      n
-                  in
-                  let v = `Node n in
-                  (aux [@tailcall]) ((step, v) :: acc) rest)
+                  aux ((step, v) :: acc) rest
```

Comment:
Is it for perf that you don't add `[@tail]` on `aux` calls?

Reply:
no good reason to remove it, I'll add it back!